src/webgpu/web_platform/copyToTexture/canvas.spec.ts

```diff
@@ -51,53 +38,137 @@ class F extends CopyToTextureUtils {
+      ctx.fillStyle = 'rgba(255, 255, 255, 0.0)';
+    } else {
+      ctx.fillStyle = 'rgba(0, 0, 0,' + alphaValue + ')';
     }
+    ctx.fillRect(rectWidth, rectHeight, width - rectWidth, height - rectHeight);
+
+    return { canvas, canvasContext };
+  }
+
+  initGLCanvasContent({
+    canvasType,
+    contextName,
+    width,
+    height,
+    premultiplied,
+    isLosePrecisionDstFormat,
+  }: {
+    canvasType: 'onscreen' | 'offscreen';
+    contextName: 'webgl' | 'webgl2';
+    width: number;
+    height: number;
+    premultiplied: boolean;
+    isLosePrecisionDstFormat: boolean;
+  }): {
+    canvas: HTMLCanvasElement | OffscreenCanvas;
+    canvasContext: WebGLRenderingContext | WebGL2RenderingContext;
+  } {
+    const canvas = this.createCanvas(canvasType, width, height);
+    if (canvas === null) {
+      this.skip('Cannot create canvas');
+    }
+
+    let canvasContext = null;
+    canvasContext = canvas.getContext(contextName, { premultipliedAlpha: premultiplied }) as
+      | WebGLRenderingContext
+      | WebGL2RenderingContext
+      | null;
+
+    if (canvasContext === null) {
+      this.skip(canvasType + ' canvas context not available');
+    }
+
+    const rectWidth = Math.floor(width / 2);
+    const rectHeight = Math.floor(height / 2);
+
+    const gl = canvasContext as WebGLRenderingContext | WebGL2RenderingContext;
+
+    // The rgb10a2unorm dst texture will have tiny errors when we compare actual and expectation.
+    // This is due to the convert from 8-bit to 10-bit combined with alpha value ops. So for
+    // rgb10a2unorm dst textures, we'll set alphaValue to 0.0 to test.
+    const alphaValue = isLosePrecisionDstFormat ? 1.0 : 0.6;
+
+    // For webgl/webgl2 context canvas, if the context created with premultipliedAlpha attributes,
+    // it means that the value in drawing buffer is premultiplied or not. So we should set
+    // premultipliedAlpha value for premultipliedAlpha true gl context and unpremultipliedAlpha value
+    // for the premulitpliedAlpha false gl context.
+    gl.enable(gl.SCISSOR_TEST);
+    gl.scissor(0, 0, rectWidth, rectHeight);
+    premultiplied
+      ? gl.clearColor(alphaValue, 0.0, 0.0, alphaValue)
+      : gl.clearColor(1.0, 0.0, 0.0, alphaValue);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+
+    gl.scissor(rectWidth, 0, width - rectWidth, rectHeight);
+    premultiplied
+      ? gl.clearColor(0.0, alphaValue, 0.0, alphaValue)
+      : gl.clearColor(0.0, 1.0, 0.0, alphaValue);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+
+    gl.scissor(0, rectHeight, rectWidth, height - rectHeight);
+    premultiplied
+      ? gl.clearColor(0.0, 0.0, alphaValue, alphaValue)
+      : gl.clearColor(0.0, 0.0, 1.0, alphaValue);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+
+    gl.scissor(rectWidth, rectHeight, width - rectWidth, height - rectHeight);
+    premultiplied
+      ? gl.clearColor(alphaValue, alphaValue, alphaValue, alphaValue)
+      : gl.clearColor(1.0, 1.0, 1.0, alphaValue);
+    gl.clear(gl.COLOR_BUFFER_BIT);
```

Comment:
Test description says this is black, but it's white

Reply:
Ah, yes, and it reminds me that I'm not updating test description.